src/promql/src/planner.rs

```diff
@@ -1351,6 +1377,107 @@ impl PromPlanner {
+                    name: modifier.card.clone(),
+                },
+            );
+            // apply label modifier
+            if let Some(matching) = &modifier.matching {
+                match matching {
+                    // keeps columns mentioned in `on`
+                    LabelModifier::Include(on) => {
+                        let mask = on.labels.iter().cloned().collect::<HashSet<_>>();
+                        left_tag_col_set = left_tag_col_set.intersection(&mask).cloned().collect();
+                        right_tag_col_set =
+                            right_tag_col_set.intersection(&mask).cloned().collect();
+                    }
+                    // removes columns memtioned in `ignoring`
+                    LabelModifier::Exclude(ignoring) => {
+                        // doesn't check existence of label
+                        for label in &ignoring.labels {
+                            let _ = left_tag_col_set.remove(label);
+                            let _ = right_tag_col_set.remove(label);
+                        }
+                    }
+                }
+            }
+        }
+        // ensure two sides have the same tag columns
+        if !matches!(op.id(), token::T_LOR) {
+            ensure!(
+                left_tag_col_set == right_tag_col_set,
+                CombineTableColumnMismatchSnafu {
+                    left: left_tag_col_set.into_iter().collect::<Vec<_>>(),
+                    right: right_tag_col_set.into_iter().collect::<Vec<_>>(),
+                }
+            )
+        };
+        let join_keys = left_tag_col_set
+            .into_iter()
+            .chain([self.ctx.time_index_column.clone().unwrap()])
+            .collect::<Vec<_>>();
+
+        // Generate join plan.
+        // All set operations in PromQL are "distinct"
+        match op.id() {
+            token::T_LAND => LogicalPlanBuilder::from(left)
+                .distinct()
+                .context(DataFusionPlanningSnafu)?
+                .join_detailed(
+                    right,
+                    JoinType::LeftSemi,
+                    (join_keys.clone(), join_keys),
+                    None,
+                    true,
+                )
+                .context(DataFusionPlanningSnafu)?
+                .build()
+                .context(DataFusionPlanningSnafu),
+            token::T_LUNLESS => LogicalPlanBuilder::from(left)
+                .distinct()
+                .context(DataFusionPlanningSnafu)?
+                .join_detailed(
+                    right,
+                    JoinType::LeftAnti,
+                    (join_keys.clone(), join_keys),
+                    None,
+                    true,
+                )
+                .context(DataFusionPlanningSnafu)?
+                .build()
+                .context(DataFusionPlanningSnafu),
+            token::T_LOR => {
+                // `OR` can not be expressed by `UNION` precisely.
```

Comment:
Maybe we can support the common case that uses `or` for the same schema?

Reply:
After some struggles I find it's not viable even for the same schemas 😢 Duplicate records also result to wrong data

Reply:
Do we have any plan to support it in the future in other ways?

Reply:
I'm going to make a `UnionDistinctOn` plan for it